pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1202 @@ func (o *ClusterUninstaller) validate() error {
-	logger.Debugf("Deleting IAM resources (%s)", filter)
-	defer logger.Debugf("Exiting deleting IAM resources (%s)", filter)
-	installerType := []string{"master", "worker", "bootstrap"}
-	for _, t := range installerType {
-		// Naming of IAM resources expected from https://github.com/openshift/installer as follows:
-		// $CLUSTER_NAME-master-role     $CLUSTER_NAME-worker-role    $CLUSTER_NAME-bootstrap-role
-		// $CLUSTER_NAME-master-profile  $CLUSTER_NAME-worker-profile $CLUSTER_NAME-bootstrap-profile
-		roleName := fmt.Sprintf("%s-%s-role", clusterName, t)
-		instanceProfileName := fmt.Sprintf("%s-%s-profile", clusterName, t)
-		if err := tryDeleteRoleProfileByName(roleName, instanceProfileName, session, logger); err != nil {
-			logger.Debugf("error deleting instance profile %s: %v", instanceProfileName, err)
-			return false, nil
-		}
-	}
-	return true, nil
+	logger.Info("Deleted")
+	return nil
 }
 
-// deleteInstances will find any running/pending instances that match the given filter and terminate them
-// and any instance profiles attached to the instance(s)
-func deleteInstances(session *session.Session, filter Filter, clusterName string, logger logrus.FieldLogger) (bool, error) {
-	logger.Debugf("Deleting instances (%s)", filter)
-	defer logger.Debugf("Exiting deleting instances (%s)", filter)
-
-	ec2Client := getEC2Client(session)
-	iamClient := iam.New(session)
-
-	describeInstancesInput := ec2.DescribeInstancesInput{}
-	describeInstancesInput.Filters = createEC2Filters(filter)
-
-	// only fetch instances in 'running|pending' state since 'terminated' ones take a while to really get cleaned up
-	describeInstancesInput.Filters = append(describeInstancesInput.Filters, &ec2.Filter{
-		Name:   aws.String("instance-state-name"),
-		Values: []*string{aws.String("running"), aws.String("pending")},
-	})
-
-	instancesFound := false
-	err := ec2Client.DescribeInstancesPages(&describeInstancesInput, func(results *ec2.DescribeInstancesOutput, lastPage bool) bool {
-		instancesFound = instancesFound || len(results.Reservations) > 0
-		for _, reservation := range results.Reservations {
-			for _, instance := range reservation.Instances {
-				// first delete any instance profiles (they are not tagged)
-				if instance.IamInstanceProfile != nil {
-					err := deleteInstanceProfile(instance.IamInstanceProfile.Id, iamClient, logger)
-					if err != nil {
-						logger.Debugf("error deleting instance profile: %v", err)
-						continue
-					}
-				}
-
-				// now delete the instance
-				logger.Debugf("deleting instance: %v", *instance.InstanceId)
-				_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
-					InstanceIds: []*string{instance.InstanceId},
-				})
-				if err != nil {
-					logger.Debugf("error deleting instance: %v", err)
-					continue
-				} else {
-					logger.WithField("id", *instance.InstanceId).Info("Deleted instance")
-				}
-			}
-		}
-
-		return lastPage
```

Comment:
This line (and a number like it throughout this package) were buggy.  I'd added the first instance of this bug in openshift/hive@d4b97563, but [the docs][1] say "To stop iterating, return false from the fn function."  So this pull request fixes buggy, only-request-one-page handling for these resource types:

```console
$ git cat-file -p b8538961a8de^:pkg/destroy/aws/aws.go | grep 'Pages(\|lastPage'
	if err := elbv2Client.DescribeLoadBalancersPages(&elbv2.DescribeLoadBalancersInput{}, func(results *elbv2.DescribeLoadBalancersOutput, lastPage bool) bool {
		return lastPage
	if err := elbv2Client.DescribeTargetGroupsPages(&elbv2.DescribeTargetGroupsInput{}, func(results *elbv2.DescribeTargetGroupsOutput, lastPage bool) bool {
		return lastPage
	err := ec2Client.DescribeInstancesPages(&describeInstancesInput, func(results *ec2.DescribeInstancesOutput, lastPage bool) bool {
		return lastPage
```

[1]: https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeInstancesPages